### PR TITLE
samples: Add C++ implementation of digits.py

### PR DESCRIPTION
**Pull request by allanrodriguez**
Currently, only a Python implementation of the SVM and KNearest digit recognition sample exists. This pull request adds a C++ implementation.

### CODE REVIEW

File: samples/cpp/digits.cpp [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+        for (int j = 0; j < M.cols; j++)
+        {
+            M.at<float>(i, j) = M_vals[i][j];
+        }
+    }
+
+    warpAffine(img, deskewed_img, M, Size(SZ, SZ), WARP_INVERSE_MAP | INTER_LINEAR);
+}
+
+class StatModel
```
Comment (berak) [inline samples/cpp/digits.cpp]: isn't this redundant (and misleading) ? 
there's already a `ml::StatModel`, a `ml::KNearest` and a `ml::SVM`

File: samples/cpp/digits.cpp [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+        }
+
+        static Ptr<KNearest> create(int k = 3)
+        {
+            return makePtr<KNearest>(k);
+        }
+
+        void load(const char* fn)
+        {
+            _model->load<ml::KNearest>(fn);
```
Comment (berak) [inline samples/cpp/digits.cpp]: this won't work (same for SVM).
`Algorithm::load()` is a static method, and returns a new instance, 

[see here for a usage example](https://github.com/opencv/opencv/blob/master/samples/cpp/letter_recog.cpp#L88)

File: samples/cpp/digits.cpp [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+
+    vector<Mat> digits;
+    vector<int> labels;
+
+    load_digits(DIGITS_FN, digits, labels);
+
+    cout << "preprocessing..." << endl;
+
+    // shuffle digits
+    shuffle(digits, labels);
```
Comment (berak) [inline samples/cpp/digits.cpp]: not sure about KNearest, but the SVM will do internal shuffeling already

File: samples/cpp/digits.cpp [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+        double sum = 0;
+
+        for (size_t i = 0; i < hist.size(); i++)
+        {
+            sum += hist[i];
+        }
+
+        for (size_t i = 0; i < hist.size(); i++)
+        {
+            hist[i] /= sum + eps;
```
Comment (berak) [inline samples/cpp/digits.cpp]: (sum+eps);   // '+' binds weaker than '/='

File: samples/cpp/digits.cpp (hunk 1 above)
Comment (allanrodriguez) [reply samples/cpp/digits.cpp]: I copied the Python implementation of this sample, which wraps all three classes, but I agree that it's redundant.

**Pull request merged**